x/stake/keeper/validator.go

```diff
@@ -246,13 +254,45 @@ func (k Keeper) UpdateValidator(ctx sdk.Context, validator types.Validator) type
 			bz := k.cdc.MustMarshalBinary(validator.ABCIValidator())
 			store.Set(GetTendermintUpdatesKey(validator.Owner), bz)
 		}
-
 	}
 
 	k.SetValidator(ctx, validator)
 	return validator
 }
 
+// updateCliffValidator determines if the current cliff validator needs to be
+// updated or swapped.
+func (k Keeper) updateCliffValidator(ctx sdk.Context, affectedVal types.Validator) {
+	cliffAddr := sdk.AccAddress(k.GetCliffValidator(ctx))
+	if !bytes.Equal(cliffAddr, affectedVal.Owner) {
+		return
+	}
+
+	pool := k.GetPool(ctx)
+
+	// NOTE: We get the power via affectedVal since the store (by power key)
+	// has yet to be updated.
+	affectedValPower := affectedVal.GetPower()
+
+	// get all the bonded validators ranked by power in descending order
+	bondedVals := k.GetValidatorsByPower(ctx)
```

Comment:
This is correct but inefficient - we shouldn't need to iterate over the bonded validator set again. Since we know what the old cliff validator power key was, we can iterate forwards (in order of increasing power) from that key to find the  next-in-line `newCliffVal`

I think that's worth doing since much of the benefit of the cliff validator optimization is to make `UpdateBondedValidators` *not* require iteration over all 100 bonded validators.

Reply:
Yeah that was a concern of mine. I'll update it to start at the cliff power key so we can make this more efficient. Thanks 👍